File: problems-extend/3201-3300/3294_convert-doubly-linked-list-to-array-ii/solution_backward.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn to_array(node: Option<Rc<RefCell<DoublyListNode>>>) -> Vec<i32> {
        // Walk `next` to the tail without collecting anything; the
        // backward sweep over `prev` then gathers the whole list, tail
        // first. One in-place reverse turns that tail-to-head buffer into
        // the answer.
        let mut current = node;
        while let Some(next) = current.as_ref().and_then(|node| node.borrow().next.clone()) {
            current = Some(next);
        }
        let mut values = Vec::new();
        while let Some(node) = current {
            values.push(node.borrow().val);
            current = node.borrow().prev.clone();
        }
        values.reverse();
        values
    }
}

File: problems-originals/3201-3300/3294_convert-doubly-linked-list-to-array-ii/solution_backward.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public int[] toArray(DoublyListNode node) {
        // Walk `next` to the tail without collecting anything; the backward
        // sweep over `prev` then gathers the whole list, tail first. One
        // mirrored pass writes that tail-to-head sequence into the result
        // array back to front.
        while (node.next != null) {
            node = node.next;
        }
        List<Integer> backward = new ArrayList<>();
        for (; node != null; node = node.prev) {
            backward.add(node.val);
        }
        int[] values = new int[backward.size()];
        for (int index = 0; index < values.length; index++) {
            values[values.length - 1 - index] = backward.get(index);
        }
        return values;
    }
}
